Web Inspector: Elements: add support for `@starting-style`

dcrousso opened the pull request:
#### ee26e01ced88303f346791a2ef7b0aa1bdfd0b90
<pre>
Web Inspector: Elements: add support for `@starting-style`
<a href="https://bugs.webkit.org/show_bug.cgi?id=294425">https://bugs.webkit.org/show_bug.cgi?id=294425</a>

Reviewed by Tim Nguyen.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
(WebCore::atRuleIdentifierForType):
(WebCore::protocolGroupingTypeForStyleRuleType):
Add a new enum value for `@starting-style`.

* Source/WebCore/style/InspectorCSSOMWrappers.cpp:
(WebCore::Style::InspectorCSSOMWrappers::collect):
Iterate over all rules within `@starting-style` when gathering information for Web Inspector.

* Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
(WI.CSSGrouping.prototype.get editable):
(WI.CSSGrouping.prototype.get isStartingStyle): Added.
(WI.CSSGrouping.prototype.get prefix):
(WI.CSSGrouping.prototype.get isSupports): Deleted.
(WI.CSSGrouping.prototype.get isLayer): Deleted.
(WI.CSSGrouping.prototype.get isContainer): Deleted.
(WI.CSSGrouping.prototype.get isStyle): Deleted.
Remove unused methods and simplify the remainder.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._markOverriddenProperties):
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.get isStartingStyle): Added.
Properties in `@starting-style` will always be overridden by anything else that matches, regardless of the cascade order/`!important`.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._renderGroupings):
Don&apos;t try to make the grouping text editable if there isn&apos;t any (e.g. `@starting-style` vs `@media foo`).

* LayoutTests/inspector/css/getMatchedStylesForNode.html:
* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNodeStartingStyleGrouping.html: Added.
* LayoutTests/inspector/css/getMatchedStylesForNodeStartingStyleGrouping-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/296925@main">https://commits.webkit.org/296925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc6307a0caf15f0eeb7f0bb1b7bc7d51a7b46fd9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29653 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20085 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116016 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111957 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38240 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83623 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24186 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64065 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23558 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/17206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59811 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/102485 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93559 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118807 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108547 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37034 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27454 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92597 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95324 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92421 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23549 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37407 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15145 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32916 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36928 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42399 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132822 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36590 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35936 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39930 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38299 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AnujPanta1 commented on LayoutTests/inspector/css/getMatchedStylesForNodeStartingStyleGrouping-expected.txt:
Just curious how do you go about generating these expected files ? Are they manually written or is there a script to generate the expected output ?

dcrousso replied:
you can use `Tools/Scripts/run-webkit-tests --reset-results path/to/test.html` to (re)generate these expected result files

nt1m commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -922,10 +922,28 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
-                        continue;
+                    } else {
+                        let propertyIsLessImportant = property.ownerStyle.isStartingStyle;
+                        let effectivePropertyIsLessImportant = effectiveProperty.ownerStyle.isStartingStyle;
```
Maybe this is just me because I understand how `@starting-style` works, but I'd name this `(effective)propertyIsStartingStyle` because it would make the code below make a bit more sense when you read it.

Even if the current name was made more generic to account for future at-rules, it's not guaranteed future at-rules re-use this exact same logic (and it you might need to define the importance of different at-rules relative to each other), so I'd rather be explicit with this being about `@starting-style`

dcrousso replied:
yeah i suppose there's no reason to have this be broader than it needs to be right now

nt1m commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -922,10 +922,28 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
+                        let propertyIsLessImportant = property.ownerStyle.isStartingStyle;
+                        let effectivePropertyIsLessImportant = effectiveProperty.ownerStyle.isStartingStyle;
+
+                        if (effectivePropertyIsLessImportant && !propertyIsLessImportant && !property.anonymous) {
```
What does `property.anonymous` mean? I'm curious what's an anonymous property in this case.

dcrousso replied:
this typically represents longhand properties that aren't actually present in the original source

for example
```css
margin: 10px;
```
will have the following `anonymous` properties behind the scenes
```css
margin-top: 10px;
margin-right: 10px;
margin-bottom: 10px;
margin-left: 10px;
```

nt1m replied:
maybe a good follow up would be renaming this `isExpandedLonghand` or something more specific

nt1m replied:
I'm not sure I understand why we're excluding them in this specific check, but I trust your judgment

dcrousso committed:
Web Inspector: Elements: add support for `@starting-style`
https://bugs.webkit.org/show_bug.cgi?id=294425

Reviewed by Tim Nguyen.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
(WebCore::atRuleIdentifierForType):
(WebCore::protocolGroupingTypeForStyleRuleType):
Add a new enum value for `@starting-style`.

* Source/WebCore/style/InspectorCSSOMWrappers.cpp:
(WebCore::Style::InspectorCSSOMWrappers::collect):
Iterate over all rules within `@starting-style` when gathering information for Web Inspector.

* Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
(WI.CSSGrouping.prototype.get editable):
(WI.CSSGrouping.prototype.get isStartingStyle): Added.
(WI.CSSGrouping.prototype.get prefix):
(WI.CSSGrouping.prototype.get isSupports): Deleted.
(WI.CSSGrouping.prototype.get isLayer): Deleted.
(WI.CSSGrouping.prototype.get isContainer): Deleted.
(WI.CSSGrouping.prototype.get isStyle): Deleted.
Remove unused methods and simplify the remainder.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._markOverriddenProperties):
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.get isStartingStyle): Added.
Properties in `@starting-style` will always be overridden by anything else that matches, regardless of the cascade order/`!important`.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._renderGroupings):
Don't try to make the grouping text editable if there isn't any (e.g. `@starting-style` vs `@media foo`).

* LayoutTests/inspector/css/getMatchedStylesForNode.html:
* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNodeStartingStyleGrouping.html: Added.
* LayoutTests/inspector/css/getMatchedStylesForNodeStartingStyleGrouping-expected.txt: Added.

Canonical link: https://commits.webkit.org/296925@main
